display_list/display_list_color_source.h

```diff
@@ -0,0 +1,668 @@
+class DlLinearGradientColorSource;
+class DlRadialGradientColorSource;
+class DlConicalGradientColorSource;
+class DlSweepGradientColorSource;
+class DlUnknownColorSource;
+
+enum class DlColorSourceType {
+  kColor,
+  kImage,
+  kLinearGradient,
+  kRadialGradient,
+  kConicalGradient,
+  kSweepGradient,
+  kUnknown
+};
+
+enum class DlTileMode {
+  // Replicate the edge color if the shader draws outside of its
+  // original bounds.
+  kClamp,
+
+  // Repeat the shader's image horizontally and vertically.
+  kRepeat,
+
+  // Repeat the shader's image horizontally and vertically, alternating
+  // mirror images so that adjacent images always seam.
+  kMirror,
+
+  // Only draw within the original domain, return transparent-black everywhere
+  // else.
+  kDecal,
+};
+
+class DlColorSource
+    : public DlAttribute<DlColorSource, SkShader, DlColorSourceType> {
+ public:
+  // Return a shared_ptr holding a DlColorSource representing the indicated
+  // Skia SkShader pointer.
+  //
+  // This method can detect each of the 4 recognized types from an analogous
+  // SkShader.
+  static std::shared_ptr<DlColorSource> From(SkShader* sk_filter);
+
+  // Return a shared_ptr holding a DlColorFilter representing the indicated
+  // Skia SkShader pointer.
+  //
+  // This method can detect each of the 4 recognized types from an analogous
+  // SkShader.
+  static std::shared_ptr<DlColorSource> From(sk_sp<SkShader> sk_filter) {
+    return From(sk_filter.get());
+  }
+
+  static std::shared_ptr<DlColorSource> MakeLinear(
+      const SkPoint p0,
+      const SkPoint p1,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  static std::shared_ptr<DlColorSource> MakeRadial(
+      SkPoint center,
+      SkScalar radius,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  static std::shared_ptr<DlColorSource> MakeConical(
+      SkPoint start_center,
+      SkScalar start_radius,
+      SkPoint end_center,
+      SkScalar end_radius,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  static std::shared_ptr<DlColorSource> MakeSweep(
+      SkPoint center,
+      SkScalar start,
+      SkScalar end,
+      uint32_t stop_count,
+      const uint32_t* colors,
+      const float* stops,
+      DlTileMode tile_mode,
+      const SkMatrix* matrix = nullptr);
+
+  virtual bool is_opaque() const = 0;
+
+  virtual std::shared_ptr<DlColorSource> with_sampling(
+      const SkSamplingOptions& options) const {
+    return shared();
+  }
+
+  // Return a DlColorColorSource pointer to this object iff it is an Color
+  // type of ColorSource, otherwise return nullptr.
+  virtual const DlColorColorSource* asColor() const { return nullptr; }
```

Comment:
Why not pure virtual?

Reply:
So that the various subclasses only need to implement the one that applies to their type?

Is there an advantage to pure virtual?

The unit tests very explicitly (and somewhat thoroughly) test the return values of these methods, so maintenance or forcing intentional implementation shouldn't be an issue.

Reply:
Sounds good to me.